src/models/src/entity/atproto.rs

```diff
@@ -0,0 +1,368 @@
+                    },
+                ..
+            }) => (did, email, email_confirmed),
+            Err(error) => {
+                error!(%error, "failed to get session for atproto");
+
+                return Err(ErrorResponse::new(
+                    ErrorResponseType::Internal,
+                    "failed to get session for atproto",
+                ));
+            }
+        };
+
+        let link_cookie = ApiCookie::from_req(req, PROVIDER_LINK_COOKIE)
+            .and_then(|value| AuthProviderLinkCookie::try_from(value.as_str()).ok());
+
+        if email.is_none() {
+            todo!()
+        }
+
+        let claims_user_id = did.to_string();
+
+        let user_opt = match User::find_by_federation("atproto", &claims_user_id).await {
+            Ok(user) => {
+                debug!(
+                    "found already existing user by federation lookup: {:?}",
+                    user
+                );
+                Some(user)
+            }
+            Err(_) => {
+                debug!("did not find already existing user by federation lookup - making sure email does not exist");
+
+                if let Ok(mut user) = User::find_by_email(email.as_ref().unwrap().to_string()).await
+                {
+                    if let Some(link) = link_cookie.as_ref() {
+                        if link.provider_id != "atproto" {
+                            return Err(ErrorResponse::new(
+                                ErrorResponseType::BadRequest,
+                                "bad provider_id in link cookie".to_string(),
+                            ));
+                        }
+
+                        if link.user_id != user.id {
+                            return Err(ErrorResponse::new(
+                                ErrorResponseType::BadRequest,
+                                "bad user_id in link cookie".to_string(),
+                            ));
+                        }
+
+                        if link.user_email != user.email {
+                            return Err(ErrorResponse::new(
+                                ErrorResponseType::BadRequest,
+                                "Invalid E-Mail".to_string(),
+                            ));
+                        }
+
+                        user.auth_provider_id = Some("atproto".to_owned());
+                        user.federation_uid = Some(claims_user_id.clone());
+
+                        Some(user)
+                    } else {
+                        return Err(ErrorResponse::new(ErrorResponseType::Forbidden, format!(
+                            "User with email '{}' already exists but is not linked to this provider.",
+                            user.email
+                        )));
+                    }
+                } else {
+                    None
+                }
+            }
+        };
+        debug!("user_opt:\n{:?}", user_opt);
+
+        let now = OffsetDateTime::now_utc().unix_timestamp();
```

Comment:
You can't know this yet, but it would be nice if you use `chrono` instead of `time`. I am getting rid of the `time` dependency whenever I stumble over leftovers as you only need / want one of them usually.